File: leetcode/2. problem/java/343.整数拆分.java
import java.time.format.TextStyle;

/*
 * @lc app=leetcode.cn id=343 lang=java
 *
 * [343] 整数拆分
 */

// @lc code=start
class Solution {
    public int integerBreak(int n) {
        // 普通版动态规划
        // return extracted2(n);

        // 简化动态规划
        // return extracted(n);

        // test
        return test(n);
    }

    private int test(int n) {
        int[] dp = new int[n + 1];
        for (int i = 2; i <= n; i++) {
            for (int j = 00; j < i; j++) {
                dp[i] = Math.max(dp[i], Math.max(j * (i - j), dp[i - j] * j));
            }
        }
        return dp[n];
    }

    /**
     * Your runtime beats 75.02 % of java submissions
     * Your memory usage beats 60.92 % of java submissions (37.8 MB)
     * 
     * @param n
     * @return
     */
    private int extracted2(int n) {
        int[] dp = new int[n + 1];
        for (int i = 2; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                // 此处的思维略微比较难理解：注意 (i-j)*j -> 每个数至少能够被分为两个数字
                dp[i] = Math.max(dp[i], Math.max((i - j) * j, dp[i - j] * j));
            }
        }
        return dp[n];
    }

    /**
     * 
     * @param n
     * @return
     */
    private int extracted(int n) {
        // dp[i] 记录数字i的最大n个正整数和,当 n≥2 时可以拆分成至少两个正整数的和
        // dp[i] = Math.max(dp[i], Math.max((i - j) * j, dp[i - j] * j));
        int[] dp = new int[n + 1];
        for (int i = 2; i <= n; i++) {
            int curMax = 0;
            for (int j = 1; j < i; j++) {
                curMax = Math.max(curMax, Math.max(j * (i - j), j * dp[i - j]));
            }
            dp[i] = curMax;
        }
        return dp[n];
    }
}
// @lc code=end

File: leetcode/2. problem/343.整数拆分.go
/*
 * @lc app=leetcode.cn id=343 lang=golang
 *
 * [343] 整数拆分
 */
package main

// @lc code=start
func integerBreak(n int) int {
	return newFunction21(n)
}

// 50/50 cases passed (0 ms)
// Your runtime beats 100 % of golang submissions
// Your memory usage beats 48.86 % of golang submissions (1.8 MB)
func newFunction21(n int) int {
	var max = func(i, j int) int {
		if i > j {
			return i
		}
		return j
	}
	dp := make([]int, n+1)
	for i := 2; i <= n; i++ {
		for j := 0; j < i; j++ {
			dp[i] = max(dp[i], max(dp[i-j]*j, (i-j)*j))
		}
	}
	return dp[n]
}

// @lc code=end
